openbb_sdk/providers/intrinio/openbb_intrinio/models/options_chains.py

```diff
@@ -302,6 +108,219 @@ class IntrinioOptionsChainsFetcher(
+        after: str
+            Return option contract expiration dates after this date. Format: YYYY-MM-DD
+        before: str
+            Return option contract expiration dates before this date. Format: YYYY-MM-DD
+        """
+
+        url = (
+            f"https://api-v2.intrinio.com/options/expirations/{symbol}/"
+            f"eod?before={before}&after={after}&api_key={api_key}"
+        )
+        r = make_request(url)
+
+        if r.status_code != 200:
+            raise RuntimeError("HTTP Error -> Status Code: " + str(r.status_code))
+
+        expirations = pd.DatetimeIndex(sorted(list(r.json()["expirations"])))
+        return list(filter(lambda x: x > pd.to_datetime(after), expirations))
+
+    @staticmethod
+    def generate_url(symbol, expiration, date, api_key):
+        return f"https://api-v2.intrinio.com/options/chain/{symbol}/{expiration}/eod?date={date}&api_key={api_key}"
+
+    @staticmethod
+    def download_json_files(urls):
+        results = []  # List to store the downloaded JSON data
+
+        def download_single_file(url):
+            response = make_request(url)
+            json_data = response.json()
+            results.append(json_data["chain"])
+
+        with concurrent.futures.ThreadPoolExecutor() as executor:
+            executor.map(download_single_file, urls)
+
+        return results
+
+    @staticmethod
+    def get_historical_chain_with_greeks(
+        symbol: str, date: Optional[Union[str, dateType]] = "", api_key: str = ""
+    ):
+        symbol = symbol.upper()
+        SYMBOLS = IntrinioOptionsChainsFetcher.get_options_tickers(api_key)
+        if symbol not in SYMBOLS:
+            raise RuntimeError(f"{symbol}", "is not supported by Intrinio.")
+
+        if not date:
+            date = datetime.now().strftime("%Y-%m-%d")
+
+        # If the symbol is an index, it needs to be preceded with, $.
+        if symbol in TICKER_EXCEPTIONS:
+            symbol = "$" + symbol
+
+        expirations = IntrinioOptionsChainsFetcher.get_options_expirations(
+            symbol=symbol, api_key=api_key, after=date
+        )
+        urls = []
+
+        for expiration in expirations:
+            urls.append(
+                IntrinioOptionsChainsFetcher.generate_url(
+                    symbol, expiration, date, api_key
+                )
+            )
+
+        results = IntrinioOptionsChainsFetcher.download_json_files(urls)
+
+        # Error handling for when data from the current date is not available yet.
+        if len(results) == 0:
+            date = pd.to_datetime(date) + timedelta(days=-1)
+            for expiration in expirations:
+                urls.append(
+                    IntrinioOptionsChainsFetcher.generate_url(
+                        symbol, expiration, date, api_key
+                    )
+                )
+            results = IntrinioOptionsChainsFetcher.download_json_files(urls)
+
+        df = pd.DataFrame()
+
+        for i in range(0, len(results)):
```

Comment:
This function has n^2 complexity and can be optimized. Hardcoding can be avoided.